contracts/contracts/oracle/OETHOracleUpdater.sol

```diff
@@ -0,0 +1,131 @@
+    /// The vault charges a 0.5% withdraw fee and the oracle prices of
+    /// the vault collateral assets are capped at 1 so the max price is 0.995.
+    /// @dev A new OETHOracleUpdater needs to be deployed if the vault withdraw fee changes.
+    uint256 public constant MAX_VAULT_PRICE = 995e15;
+
+    /// @notice The OETH/ETH Curve pool
+    ICurvePool public immutable curvePool;
+    /// @notice The OETH Vault
+    IVault public immutable vault;
+
+    struct OracleUpdaterConfig {
+        address vault;
+        address curvePool;
+    }
+
+    event AddPrice(
+        address indexed oracle,
+        uint256 answer,
+        uint256 vaultPrice,
+        uint256 marketPrice
+    );
+
+    /**
+     * @param _vault Address of the OETH Vault
+     * @param _curvePool Address of the OETH/ETH Curve pool
+     */
+    constructor(address _vault, address _curvePool) Governable() {
+        curvePool = ICurvePool(_curvePool);
+        vault = IVault(_vault);
+    }
+
+    /// @notice Adds a new on-chain, aggregated OETH/ETH price to 18 decimals to the specified Oracle.
+    /// @dev Callable by anyone as the prices are sourced and aggregated on-chain.
+    /// @param oracle Address of the Oracle that has authorized this contract to add prices.
+    function addPrice(IOracleReceiver oracle) external {
+        (
+            uint256 answer,
+            uint256 vaultPrice,
+            uint256 marketPrice
+        ) = _getPrices();
+
+        emit AddPrice(address(oracle), answer, vaultPrice, marketPrice);
+
+        // Add the new aggregated price to the oracle.
+        // Authorization is handled on Oracle side
+        oracle.addPrice(uint128(answer));
+    }
+
+    /**
+     * @dev Gets the OETH/ETH market price, vault floor price and aggregates to a OETH/ETH price to 18 decimals.
+     */
+    function _getPrices()
+        internal
+        view
+        returns (
+            uint256 answer,
+            uint256 vaultPrice,
+            uint256 marketPrice
+        )
+    {
+        // Get the aggregated market price from on-chain DEXs
+        marketPrice = _getMarketPrice();
+
+        // If market price is equal or above the vault price with the withdraw fee
+        if (marketPrice >= MAX_VAULT_PRICE) {
+            answer = marketPrice;
+            // Avoid getting the vault price as this is gas intensive.
+            // its not going to be higher than 0.995 with a 0.5% withdraw fee
+            vaultPrice = MAX_VAULT_PRICE;
+        } else {
+            // Get the price from the Vault. This includes the withdraw fee
+            // and the vault collateral assets priced using oracles
+            vaultPrice = vault.floorPrice();
+
+            if (marketPrice > vaultPrice) {
+                // Return the market price with the Vault price as the floor price
+                answer = marketPrice;
+            } else {
+                // Return the vault price
+                answer = vaultPrice;
```

Comment:
And I think we've said that if market price is lower than the vaultPrice, then we might go with the market price.

Thinking more about this, I am not so sure this is a good idea, since the market price can be manipulated downwards even though there is a smaller financial incentive to do it. The way it is in the codebase right now seems safer to me.

probably something we should discuss more about.